### DISCUSSION TITLE
support-complex

### DISCUSSION THREAD

**Discussion by giscus** [mrocklin/mrocklin.github.io#25] [Giscus]
# support-complex

Supporting a Python product is difficult because it’s hard to draw a line between your product and the rest of the ecosystem. This creates confusion, especially when it comes to triage and attribut...

https://matthewrocklin.com/support-complex.html

**Comment by alohabrad** [comment]
This post does a good job capturing the challenges of supporting an open ecosystem like Python. I wonder if there’s also an opportunity here: Could a product bridge some of these gaps, making the open ecosystem feel more similar to a closed, vertically integrated solution while still being open? Would people pay for that kind of "open simplicity"?

**Reply by mrocklin** [reply]
I don't know that there's a general technology solution to this kind of problem (unless maybe you count LLMs)

Instead, I think that we see three ways in which this gets handled in the market today:

1. Professional services firms
2. Eventually product companies see enough customers with the same patterns of problems that they eventually take on more and more of the scope of that pattern 
3. Some larger scoped product companies sell broad support.  Anaconda and IBM come to mind for Python.  My experience is that the support service they provide degrades as it becomes more general

Maybe though you have some idea that you're trying to hint towards?

**Reply by mrocklin** [reply]
Another analogy might be in home construction, where people use general contractors to manage all sorts of different specialists (electricians, plumbers, framers, etc..)  General wisdom is that no, there isn't a technology that makes this problem smooth.  Instead it's just lots of custom last-mile engineering on the part of the general contractor (hat tip to Matt Plough for this analogy).

Maybe what I'm hearing you propose is like the growth in manufactured prefabricated homes (not like a double-wide, but like factory-built pieces that are assembled on-site).  Cool technology and I hope that it takes off.  Probably that's mostly a reinvention of the stack though.

**Reply by alohabrad** [reply]
Great analogy. The persona in the Support Request section feels like a DIY homebuilder — learning plumbing via YouTube. Anaconda et al. are the general contractors many don’t want to use or can’t afford. If the core issue is that supporting a Python product is hard and expensive because it’s tough to draw a line between your product and the rest of the ecosystem, solving that seems necessary for the Python ecosystem to scale.

The prefab home analogy feels like the opportunity: a technology that helps the Python ecosystem just work for the DIY builder. Scalable, automated broad support — like what professional services firms offer — but delivered through software for the long tail of customers (maybe powered by LLMs, as you say).
